CI errors look related:

```
../scipy/sparse/linalg/_eigen/arpack/_arpackmodule.c:14:13: error: 'ARNAUD_cplx' macro redefined [-Werror,-Wmacro-redefined]
   14 |     #define ARNAUD_cplx(real, imag) ((real) + (imag)*I)
      |             ^
../scipy/sparse/linalg/_eigen/arpack/arnaud/include/arnaud/types.h:27:13: note: previous definition is here
   27 |     #define ARNAUD_cplx(real, imag) CMPLX(real, imag)
      |             ^
../scipy/sparse/linalg/_eigen/arpack/_arpackmodule.c:15:13: error: 'ARNAUD_cplxf' macro redefined [-Werror,-Wmacro-redefined]
   15 |     #define ARNAUD_cplxf(real, imag) ((real) + (imag)*I)
      |             ^
../scipy/sparse/linalg/_eigen/arpack/arnaud/include/arnaud/types.h:26:13: note: previous definition is here
   26 |     #define ARNAUD_cplxf(real, imag) CMPLXF(real, imag)
```

--- a/scipy/sparse/linalg/_eigen/arpack/_arpackmodule.c
+++ b/scipy/sparse/linalg/_eigen/arpack/_arpackmodule.c
@@ -7,13 +7,8 @@
 #include "numpy/arrayobject.h"
 #include "arnaud/include/arnaud/arnaud.h"
 
-#if defined(_MSC_VER)
-    #define ARNAUD_cplx(real, imag) ((_Dcomplex){real, imag})
-    #define ARNAUD_cplxf(real, imag) ((_Fcomplex){real, imag})
-#else
-    #define ARNAUD_cplx(real, imag) ((real) + (imag)*I)
-    #define ARNAUD_cplxf(real, imag) ((real) + (imag)*I)
-#endif
+/* ARNAUD_cplx/ARNAUD_cplxf are now defined in arnaud/types.h
+   (included via arnaud.h) using CMPLX/CMPLXF to handle NaN correctly. */
 
 #ifdef HAVE_BLAS_ILP64
     #define ARNAUD_PyLong_As(obj)   (ARNAUD_INT)PyLong_AsLongLong(obj)

--- a/scipy/sparse/linalg/_eigen/arpack/arnaud/include/arnaud/types.h
+++ b/scipy/sparse/linalg/_eigen/arpack/arnaud/include/arnaud/types.h
@@ -2,7 +2,48 @@
 #define ARNAUD_TYPES_H
 
 #include <stdint.h>
-#include "scipy_complex_support.h"
+#include <complex.h>
+
+/*
+ * Portable CMPLX / CMPLXF fallbacks.
+ *
+ * The C11 standard provides CMPLX/CMPLXF, but not all compilers expose
+ * them (notably MSVC).  The  real + imag * I  pattern corrupts the real
+ * part when imag is NaN.
+ *
+ * Fallback priority:
+ *   1. Compiler-provided CMPLX / CMPLXF  (C11, already defined)
+ *   2. MSVC intrinsics  (_Cbuild / _FCbuild)
+ *   3. __builtin_complex  (GCC >= 4.7, Clang)
+ *   4. Union type-pun     (portable last resort)
+ */
+#ifndef CMPLX
+    #if defined(_MSC_VER)
+        #define CMPLX(x, y) _Cbuild(x, y)
+    #elif defined(__has_builtin)
+        #if __has_builtin(__builtin_complex)
+            #define CMPLX(x, y) __builtin_complex((double)(x), (double)(y))
+        #endif
+    #endif
+    #ifndef CMPLX
+        #define CMPLX(x, y) \
+            ((union { double a[2]; double complex z; }){{(x), (y)}}).z
+    #endif
+#endif
+
+#ifndef CMPLXF
+    #if defined(_MSC_VER)
+        #define CMPLXF(x, y) _FCbuild(x, y)
+    #elif defined(__has_builtin)
+        #if __has_builtin(__builtin_complex)
+            #define CMPLXF(x, y) __builtin_complex((float)(x), (float)(y))
+        #endif
+    #endif
+    #ifndef CMPLXF
+        #define CMPLXF(x, y) \
+            ((union { float a[2]; float complex z; }){{(x), (y)}}).z
+    #endif
+#endif
 
 
 /* Integer type for BLAS/LAPACK interface */

